### ISSUE TITLE
Cache may contain stale data

### ISSUE THREAD

**Issue by glommer** [scylladb/scylla#1364]
Scylla version (or git commit hash): origin/master

We currently have a problem in our cache management. We don't have a reproducer for this yet, but here is how it goes:

(From an e-mail from me) 

> There is a set of sstables S, and two sstables s1 and s2 are currently being flushed. Say s1 is a very big table that was flushed first because it reached maximum size, and s2 is a small table that   was flushed manually. Because of that, all timestamps in s1 are older than s2.

> Now assume both s1 and s2  contain a mutation for partition p_i.

> s1 finishes first, and start the update process that can take a long time. At this point, it uses the set S in the presence checker. 

> s2 then finishes, but because it is very small the update process is fast. It uses the set { S + s1 } in  the presence checker - specially since s1 adds itself to the sstable list before calling update_cache. It checks to see if p_i is in the cache, it isn't yet because s1 hasn't reached that point yet. Then if goes to the presence checker to check existence of p_i, but it exists in s1 - so we get a maybe exists, and nothing is done.

> Now s1 finally reaches the point where it will deal with pi. It doesn't know s2 is flushed because it is  using the old set S of sstables. Data for p_i is is not in the cache. We check the presence checker on S, it definitely doesn't exist, and we move p_i's data into the cache. But that's stale data. We have assumed in the beginning that s1 is an older sstable, and the timestamps of its mutations are older.

> It seems to me that this sequence of events could lead to the wrong data being present in the cache, while the new one is not accessible.

**Comment by glommer** [comment]
@tgrabiec says it is easy to fix. Would you like to get this one, Tomek ?

**Comment by slivne** [comment]
@elcallio - is it ok to finish flushing a memtable with a higher RP
position then finishing flushing the previous memtable with a lower RP -
what happens if we crash - in between - will we try to replay the info ?

From the limited understand I have we are supposed to make sure that we
finish flushing "in-order", if we must flush in-order then the issue may be
that we allowed s2 to finish flushing ?

**Comment by glommer** [comment]
We flush in order, but nothing guarantees that :

1) the flushes will end in order
2) the update process will end in order. The update process itself is CPU
intensive, and because of that we split it into multiple continuations.

**Comment by slivne** [comment]
The simplest solution would be to not allow flushing in parallel for the
same cf for ongoing memtables (not ones related to streaming/compaction
etc.)

While in some cases the disk may be able to handle that in others we may be
delaying the flush of the first memtable by inserting additional io.

**Comment by glommer** [comment]
On Sun, Jun 19, 2016 at 2:02 AM, Shlomi Livne <notifications@github.com>
wrote:

> The simplest solution would be to not allow flushing in parallel for the
> same cf for ongoing memtables (not ones related to streaming/compaction
> etc.)
>
> While in some cases the disk may be able to handle that in others we may be
> delaying the flush of the first memtable by inserting additional io.


That could make us potentially very, very slow. We need to fix this bug. It
is not that hard.


>

**Comment by elcallio** [comment]
For the commit log, in runtime, it is ok that flushes finish out-of-order, as long as the actual _reporting_ to the CL is done (via the flush-queue in database.cc) in order. This is required to prevent deleting a segment that contains unflushed data. 

However, there is a big issue with writing sstables out-of-order, and that is replay. Each sstable contains its high replay position mark. Replay looks at existing sstables and checks this, and assumes that if an sstable exists with mark A, then all data before A was also persisted. 
Now, obviously, if we write out of order, finish a _later_ sstable on disk first (so it exists and is complete), but fail to complete one with lower RP data, then this will break down. 
Origin forces actual write order (somewhat implicitly, through the use of execution workers) for sstables (per CF). 
We could of course just say "crash and replay is rare", and ignore the sstable positions, which might make replay take longer, but perhaps not overtly so... 
Also worth noting is that it seems recent changes to the sstable compaction or whatnot (at start?) does not propagate the replay position marker, thus causing us to pick up "zero" markers -> must replay everything anyway.

**Comment by tgrabiec** [comment]
I am creating a separate issue for tracking the commitlog replay thing.
